tests/python/test_basic_models.py

```diff
@@ -325,3 +325,69 @@ def validate_model(parameters):
         parameters = {'tree_method': 'hist', 'booster': 'dart',
                       'objective': 'multi:softmax'}
         validate_model(parameters)
+
+    def run_slice(self, booster):
+        from sklearn.datasets import make_classification
+        num_classes = 3
+        X, y = make_classification(n_samples=1000, n_informative=5,
+                                   n_classes=num_classes)
+        dtrain = xgb.DMatrix(data=X, label=y)
+        num_parallel_tree = 4
+        num_boost_round = 16
+        total_trees = num_parallel_tree * num_classes * num_boost_round
+        booster = xgb.train({
+            'num_parallel_tree': 4, 'subsample': 0.5, 'num_class': 3, 'booster': booster},
+                            num_boost_round=num_boost_round, dtrain=dtrain)
+        assert len(booster.get_dump()) == total_trees
+        beg = 3
+        end = 7
+        sliced: xgb.Booster = booster[beg: end]
+
+        sliced_trees = (end - beg) * num_parallel_tree * num_classes
+        assert sliced_trees == len(sliced.get_dump())
+
+        sliced_trees = sliced_trees // 2
+        sliced: xgb.Booster = booster[beg: end: 2]
+        assert sliced_trees == len(sliced.get_dump())
+
+        sliced: xgb.Booster = booster[beg: ...]
+        sliced_trees = (num_boost_round - beg) * num_parallel_tree * num_classes
+        assert sliced_trees == len(sliced.get_dump())
+
+        sliced: xgb.Booster = booster[beg:]
+        sliced_trees = (num_boost_round - beg) * num_parallel_tree * num_classes
+        assert sliced_trees == len(sliced.get_dump())
+
+        sliced: xgb.Booster = booster[:end]
+        sliced_trees = end * num_parallel_tree * num_classes
+        assert sliced_trees == len(sliced.get_dump())
+
+        sliced: xgb.Booster = booster[...:end]
+        sliced_trees = end * num_parallel_tree * num_classes
+        assert sliced_trees == len(sliced.get_dump())
+
+        self.assertRaises(ValueError, lambda: booster[-1: 0])
+        # we do not accept empty slice.
+        self.assertRaises(ValueError, lambda: booster[1:1])
+        # stop can not be smaller than begin
+        self.assertRaises(ValueError, lambda: booster[3:0])
+        self.assertRaises(ValueError, lambda: booster[3:-1])
+        # negative step is not supported.
+        self.assertRaises(ValueError, lambda: booster[0:2:-1])
+        # step can not be 0.
+        self.assertRaises(ValueError, lambda: booster[0:2:0])
+
+        trees = [_ for _ in booster]
+        assert len(trees) == num_boost_round
+
+        self.assertRaises(TypeError, lambda: booster["wrong type"])
+        self.assertRaises(IndexError, lambda: booster[:num_boost_round+1])
+        self.assertRaises(ValueError, lambda: booster[1, 2])  # too many dims
+
+        def assign():
+            booster[...:end] = booster
+        self.assertRaises(TypeError, assign)
+
+    def test_slice(self):
+        self.run_slice('gbtree')
+        self.run_slice('dart')
```

Comment:
Can we use `@pytest.mark.parameterize` instead?

```python
@pytest.mark.parameterize(booster, ['gbtree', 'dart'])
def run_slice(self, booster):
```

See examples at [Parameterizing tests](https://docs.pytest.org/en/6.1.2/example/parametrize.html)

Reply:
It seems not being compatible with class  method:
```
TypeError: test_slice() missing 1 required positional argument: 'booster'
```

Reply:
Try `@pytest.mark.parameterize('booster',['gbtree', 'dart'])` (note the quotes around `booster`

Reply:
@trivialfis Also, `TestModels` should not be a subclass of `unittest.TestCase`. https://stackoverflow.com/a/35562401. Try making the class a subclass of `object`:
```python
class TestModels(object):
```

Reply:
Thanks for the suggestion, done.